Initial structure + some definitions.

artursouza opened the pull request.

paulyuk commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+### Spec Versioning
+
+This spec follows [semantic versioning 2.0](https://semver.org/spec/v2.0.0.html). Matching the Dapr runtime API version for major and minor. It means that for each minor version release of Dapr runtime, there will be a new release of the spec. The patch version of the spec and runtime increase independently.
```
The subtlety of this is it reads (to me) as you need to enforce keeping version numbers the same between SDK and runtime.  But I believe, and correct me, this is also trying to state there needs to be 1 to 1 parity at all times between SDK and runtime.  Could we please clarify?

artursouza replied:
I reworded this to make it clear that the parity is between the spec and the runtime. The SDKs can still be versioned independently - so the table in docs that matches the SDK versions with runtime versions remains and the spec compliance is automatically done via the runtime version.

paulyuk commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+## Client Object
+
+The SDK MUST offer a way to instantiate a client object without requiring any parameter, using the `DAPR_HTTP_PORT` or `DAPR_GRPC_PORT` environment variables injected by the Dapr sidecar injector or Dapr CLI - defaulting to `3500` or `50001`, correspondingly, if environment variable is not set.
```
One more I feel passionate about is `APP_PORT` which leads to many mismatch bugs.

artursouza replied:
The SDKs don't get this environment variable injected. This is in the sidecar.

paulyuk commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+The client object MUST abstract the protocol used to communicate to the Dapr sidecar and not offer gRPC or HTTP specific input or output.
+
+The client object MIGHT offer both gRPC or HTTP implementations.
```
I notice some of our SDKs leak a number of gRPC extensions, and then the API becomes hard to use if you don't know how to use it (meaning which APIs to use and which to ignore).

I believe this guidance will really help!

msfussell replied:
Really? Can you provide an example?

artursouza replied:
I need some clarity here. I did not understand the issue.

paulyuk commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+## Actors
+
+An SDK MIGHT implement the Actors building block.
```
If we create new inter-dependencies I'd love for some ceremonious declaration of this.

artursouza replied:
Do you mean compatibilities between Actor invocations cross languages?

mukundansundar replied:
Why is this MIGHT and not MUST?  What does not `MIGHT` mean? It's not part of the initial notational convention.

artursouza replied:
The notation is defined above. I put MIGHT because Actors is kind of a separate thing from the "core APIs".

artursouza replied:
I am changing it to MUST until we decide to make it optional. It seems to be the less controversial way to go now.

msfussell commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+## Overview
+
+[Dapr](https://github.com/dapr/dapr) offers a sidecar with an HTTP and gRPC API listening on localhost. In order to offer an additional layer of convenience, Dapr also offers SDKs in a variety of languages. Additionally, actors in Dapr rely on the SDK for applications to implement their actor classes and easy inocation of actor methods, timers and reminders.
```
```suggestion
[Dapr](https://github.com/dapr/dapr) offers a sidecar with an HTTP and gRPC API listening on localhost. In order to offer an additional layer of convenience, Dapr also offers SDKs in a variety of languages. Additionally, actors in Dapr rely on the SDK for applications to implement their actor classes and innovation of actor methods, timers and reminders.
```

artursouza replied:
I meant invocation.

msfussell commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+The SDK MUST offer a way to instantiate a client object without requiring any parameter, using the `DAPR_HTTP_PORT` or `DAPR_GRPC_PORT` environment variables injected by the Dapr sidecar injector or Dapr CLI - defaulting to `3500` or `50001`, correspondingly, if environment variable is not set.
+
+The client object MUST connect to the Dapr sidecar based on the `DAPR_SIDECAR_IP` environment variable, defaulting to `127.0.0.1` if not set.
```
Does this ENV Variable exist `DAPR_SIDECAR_IP`? If so this is not used in any samples and not documented. I doubt anyone is aware of this.

artursouza replied:
It exists in the Java SDK. I can remove it from the spec until we agree to this.

msfussell commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+The client object MIGHT offer both gRPC or HTTP implementations.
+
+The client object SHOULD NOT implement any Actor functionality since that is handled as a separate construct.
```
I presume Actors are going to be covered later in this spec?

artursouza replied:
Yes. There is a separate section. And Actors will be the hardest one, IMO.

msfussell commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+The client object MIGHT provide overloaded methods with reduced number of parameters for simpler method invocations.
+
+The client object SHOULD NOT implement any of the methods for the [preview APIs](#alpha-apis)
```
```suggestion
The client object SHOULD NOT implement any of the methods for [alpha APIs](#alpha-apis)
```

msfussell commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+* [Actor client](actors/client.md)
+* [Actor runtime](actors/runtime.md)
+
+## Preview APIs
```
```suggestion
## Alpha APIs
```

msfussell commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+## Preview APIs
+
+The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object. The preview client object MUST implement the preview APIs below:
```
```suggestion
The alpha APIs MUST be offered via an alpha client object implementing a separate interface from the core's client object. The alpha client object MUST implement the alpha APIs below:
```

msfussell commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object. The preview client object MUST implement the preview APIs below:
+
+* [Configuration](preview/configuration.md)
+* [Distribured Lock](preview/distributed-lock.md)
```
```suggestion
* [Distributed Lock](alpha/distributed-lock.md)
* [State Management Query](alpha/state-management.md)
```

msfussell commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object. The preview client object MUST implement the preview APIs below:
+
+* [Configuration](preview/configuration.md)
```
```suggestion
* [Configuration](alpha/configuration.md)
```

halspang commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+The client object MIGHT provide overloaded methods with reduced number of parameters for simpler method invocations.
+
+The client object SHOULD NOT implement any of the methods for the [preview APIs](#alpha-apis)
```
I see what we're suggesting for the alpha APIs below, but I don't think this is a good solution. See my comment there but I think we should do this differently if we're able to.

artursouza replied:
I was expecting this to trigger a discussion since each SDK is doing something different here.

halspang commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+The client object SHOULD NOT implement any of the methods for the [preview APIs](#alpha-apis)
+
+## Object Serialization
```
What about accepting raw bytes? This is a pretty common ask in the .NET SDK.

artursouza replied:
OK. I will add that. We have that too. It is a special case of serialization since byte[] is usually an object.

artursouza replied:
Added that statement to make it clear.

halspang commented on spec/spec.md:
```diff
@@ -0,0 +1,109 @@
+
+## Preview APIs
+
+The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object. The preview client object MUST implement the preview APIs below:
```
I'm not a huge fan of this decision. The way we handle this in the .NET SDK is by using the `[Obsolete]` attribute. This attribute forces the caller to acknowledge that the API is not stable (compilation failure if you don't). 

Using a preview client also means that you'll end up with a breaking change when an API graduates to stable as you'll need to update your code to use the standard client. This could be avoided by having the preview client map over to the standard client but then you just have additional code paths.

artursouza replied:
This is an interesting discussion and it shows the need for a spec. I don't think we should optimize for no breaking changes for preview APIs graduating since they will break at the API level regardless. I am not sure if [obsolete] is giving the right message either since the API is not obsolete, it is "experimental".

I think the solution should make it clear that the app is using an Alpha API. Eager to see what others think here. We did this PreviewClient in the Java SDK and it looked pretty clean from the core client. It also makes it clear that it does not have the same guarantees as the core client.

halspang replied:
Yeah, I agree that `[Obsolete]` isn't the best attribute, but it's unfortunately the only one of that style in .NET. It does let you provide a message though, so when your code fails to compile it'll say something like "The ... API is current an Alpha API and is subject to change..."

The way we did it in the .NET SDK is that you won't need to have a breaking change when we change the API. The `DaprClient` is built as a layer over the gRPC provided client. So, when we update the proto, the methods will update. We can then call the new methods in the SDK and everything goes back to normal. When the user updates to the new revision, all they'll notice is that the `Obsolete` is gone but the compilation path remains the same. 

For languages that don't have support for this, it does become a bit harder. I wonder if we can do something else like an extension onto the client? You can import a separate package or project that contains those methods but they're labeled the same as the alpha APIs are an interface/inheritance/extension. When the graduate, they're moved to the base class and out of the alpha package. Thoughts?

artursouza replied:
Interesting. If they are in a different package, then it would need to be a different interface, right? I am thinking of how to make the spec language agnostic. So, asking to have a different interface will accomplish that in a way that is more language neutral. In Java SDK, it is a separate interface. The class implements both interfaces but from the user's perspective, he is getting another type when we asks for an alpha client. I don't think we should optimize not to have a breaking change in SDK when an API graduates.

rynowak replied:
> The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object.

My feedback is that this will make it too hard for users to use preview features in practice. In a complex business application, objects like a Dapr client are threaded through several layers of glue code to get to where they are actually called. This puts a lot of burden on the user to update these references to refer to the new **preview** type, and then update them again in the future to refer to the stable one.  This kind of idea can work well for dynamic type systems, but it's also less meaningful when you've got a dynamic type system.

Of the problems SDKs/Libraries face, there are two typical problems with preview features:

- Users are not aware of the preview features and thus, they don't get a lot of feedback until they are stable.
- Users accidentally use preview features and have mismatched expectations.

In my experience the former problem is way more common, so I wouldn't want to put up even more barriers.

I'd suggest being less prescriptive and talking about the goal instead. Something like:

> SDKs SHOULD use an in-code mechanism in addition to documentation to communicate to users about the instability of preview features. For a function to interface with the new **widget** building block could be called `Preview_GetWidget()`

artursouza replied:
Since this prescription seems to be controversial. I will remove it and allow us to have a separate discussion about it.

yaron2 reviewed (changes requested):
Please change all `Preview API` phrases to `Alpha API`.

artursouza commented:
> Please change all `Preview API` phrases to `Alpha API`.

Just did.

rynowak commented on spec/spec.md:
```diff
@@ -0,0 +1,111 @@
+
+The SDK MUST serialize domain objects to JSON by default, for both input and output.
+
+The SDK MUST accept `byte[]` as domain object for input and output.
```
IMO specifying the behavior of this requires a lot more thought. I'm not saying this is a bad scenario, but the expectations of what this does need to be clarifed. For example a lot of .NET JSON libraries will serialize `byte[]` as a base64-encoded JSON string. If the SDK author does the naive thing, then that would likely be the result.

rynowak replied:
My other feedback is that this is also too prescriptive. We should talk about the goal rather than the implementation. For example in .NET it's not really idiomatic to use `byte[]`. For large requests/responses this is very costly.

Someone designing an API for 'raw' I/O in .NET would probably start with `Stream` and then add `ReadOnlySpan<byte>` if there was demand for it.

mukundansundar replied:
I agree. I think should be more in terms of saying that the `SDK must accept domain objects serialized as bytes` or some such phrase rather than specifying any specific input/output format like a `byte[]`.

artursouza replied:
I can reword this statement to not force the type.

There is a good point brought by Ryan of how much prescriptive the SDK spec should be. If we agree that the differences among SDKs right now are not a concern, that might remove the need for a spec at all (IMO). At the end of this spec, there will be changes required in each SDK to be compliant with the SPEC. I know that each maintainer made decisions in isolation and (sometimes) purposely different (like in "pluggable serializer"). I am targeting a spec that is minimum and removing controversial decisions for a later point.

As the spec evolves, there will be points where SDKs would need to make changes to be "compliant" to the spec.

mukundansundar commented on spec/spec.md:
```diff
@@ -0,0 +1,111 @@
+
+#### SDK
+
+An "SDK" or "Software Development Kit" is defined as one or more libraries that offers a set of public interfaces, methods, classes and attributes offering a programming model for Dapr native to a given programming language.
```
```suggestion
An "SDK" or "Software Development Kit" is defined as one or more libraries that offers a set of public interfaces, methods, classes and attributes offering a programming model for Dapr, native to a given programming language.
```

mukundansundar commented on spec/spec.md:
```diff
@@ -0,0 +1,111 @@
+
+The SDK MUST accept `byte[]` as domain object for input and output.
+
+The SDK MIGHT accept a serializer implementation via a provided interface.
```
Should we clarify here that Actor serialization is different from other serialization and that it will use a completely different serializer?

artursouza replied:
Everything in Actors is separate.

artursouza commented on spec/core/state-store.md:
```diff
@@ -0,0 +1,49 @@
+| Name | Type | Required | Default Value | Description |
+|-----------|------|----------|---------------|-------------|
+| `storeName` | `string` | Y | | The state store name for the operation. |
+| `states` | array of `state` | N |`false` | Whether a value _must_ be provided for the parameter. |
```
Copy and paste error.

artursouza commented:
I have removed the controversial points and addressed the other comments. I think we have something to get started, so others can start contributing to the spec.
